Repair some corner cases in cancellation propagation between coroutines and listenable futures

vadimsemenov opened the pull request:
Implement bidirectional cancellation for future coroutine builder.

This also:
- Refactors JobListenableFuture infrastructure so it can be reused in CoroutineScope.future and Deferred.asListenableFuture;
- Hides ListenableFuture.asDeferred return type, so it can't be casted to CompletableDeferred;
- Adds more tests to cover fixed corner cases;
- Improves documentation;
- Suppresses annoying warnings in tests.

turansky committed:
Transition to the Gradle kts

Non-deprecated 'jmhJar' configuration (#2032) (+9 squashed commits)
Squashed commits:
[8d07d3695] Use new Kotlin/JS plugin (#1983)
* Use new Kotlin/JS plugin
* Support legacy DCE mode for 1.4-M2
[d22464030] Add Dokka configuration method
[56e1c9bfe] Dokka plugin in 'buildSrc'
[dfdd202e6] Remove unused repositories
[4cf1d02dc] Kotlin DSL - 'javafx'
[d8f7d507e] Avoid task name duplication
[f06a56b8b] Avoid task name duplication
[a09df3de5] Separate 'UnpackAar' action
[fd5bf6b9c] Separate 'RunR8' task

elizarov committed:
Fixed val counter AtomicInteger in shared state example (#2063)

Fixes #2062

0xbayo committed:
Fix word omission (#2071)

medavox committed:
In all documentation "get the full code here" (#2076)

medavox committed:
Improve grammar in docs/basics (#2076)

Inego committed:
Fix typos in docs (#2098)

Co-authored-by: Louis CAD <louis.cognault@gmail.com>

elizarov committed:
Fixed invalid self-links to index.md in docs

Fixes #2107

vadimsemenov committed:
Prevent casting of Deferred returned from ListenableFuture.asDeferred()

vadimsemenov committed:
Add a comment about potential flaw in cancellation

vadimsemenov committed:
Document the difference between awaiting on ListenableFuture and on Deferred

vadimsemenov committed:
Fix KDoc reference to Deferred.cancel

vadimsemenov committed:
Fix references to AbstractFuture

qwwdfsad committed:
Do not print error message to the console when we've failed to install signal handler. (#2118)

Otherwise, it may lead to obscure error messages in the console when users are on Windows and are not aware about signals.

vadimsemenov committed:
Add a bit more tests for ListenableFutures.

Some of them are failing (ignored or commented out) and indicate bugs in current implementation.

vadimsemenov committed:
Get rid of dependency on kotlin.Result

elizarov committed:
Fix ensureAction to work in the empty context case (#2119)

Fixes #2044

vadimsemenov committed:
Disentangle InnerFuture from the deferred.

Deferred.cancel is inherited from Job.cancel. So we can generalise the delegate to cancel.

Now InnerFuture can complete with any value, not necesserily with the passed Job.
OuterFuture completes InnerFuture with the original deferred, so this change is effectively no-op.

vadimsemenov committed:
Fix formatting error in KDoc

vadimsemenov committed:
Override afterDone instead of cancel in InnerFuture

This approach is recommended by AbstractFuture:
> Rather than override this [#cancel] method to perform additional cancellation work or cleanup,
> subclasses should override {@link #afterDone}, consulting {@link #isCancelled} and {@link
> #wasInterrupted} as necessary.

Another (yet undocumented in AbstractFuture) advantage is that afterDone is executed
before listeners.
Before this change, if call to super.cancel succeeds, it can also run all the attached listeners.
It means that if any client violated the contract and attached a heavy listener on direct executor,
deferred cancellation will be deferred (no pun intended) until after all listeners are executed.
Now, attached listeners cannot affect cancellation of deferred.

vadimsemenov committed:
Refactor OuterFuture so it doesn't depend on Deferred

The API of OuterFuture now includes:
- complete method that accepts result of a coroutine [1];
- jobToCancel [2] constructor argument to propagate cancellation to.
It's implied that the Job[2] represents the coroutine[1].
// It doesn't look like the most robust design, can it be better?

This change is a pure refactoring. It should help to reuse OuterFuture infrastructure
to implement bidirectional cancellation for CoroutineScope.future coroutine builder.

vadimsemenov committed:
Suppress unstable API warning in afterDone

vadimsemenov committed:
Implement bidirectional cancellation for future

This reuses OuterFuture infrastructure. Thus should minimise remaining gaps (if any?)
in the cancellation propagation between Coroutines and ListenableFutures.

vadimsemenov committed:
Suppress warnings in ListenableFutureTest

elizarov committed:
Merge remote-tracking branch 'origin/master' into develop

elizarov committed:
Merge remote-tracking branch 'origin/master' into develop

yorickhenning commented on integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:
```diff
@@ -33,7 +35,10 @@ import kotlin.coroutines.*
  *
  * Note that the error and cancellation semantics of [future] are _subtly different_ than
- * [asListenableFuture]'s. See [ListenableFutureCoroutine] for details.
+ * [asListenableFuture]'s.
```
Remove the linebreak here.

yorickhenning commented on integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:
```diff
@@ -17,8 +17,10 @@ import kotlin.coroutines.*
  *
- * The created coroutine is cancelled when the resulting future completes successfully, fails, or
- * is cancelled.
+ * When the created coroutine completes (either successfully or exceptionally), it will try to
```
Avoid parentheticals in docstrings, and in technical writing generally. Instead, use several shorter sentences and restate the direct object as necessary.

"When the created coroutine finishes, it *synchronously* completes the returned future with the same outcome. The completion is synchronous whether the coroutine succeeds or fails."

vadimsemenov replied:
Reworded by referencing to Job.isCancelled.

yorickhenning commented on integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:
```diff
@@ -17,8 +17,10 @@ import kotlin.coroutines.*
  *
- * The created coroutine is cancelled when the resulting future completes successfully, fails, or
- * is cancelled.
+ * When the created coroutine completes (either successfully or exceptionally), it will try to
```
Is there any way completing the Future will fail? The LF is settable/completable, so this should be able to categorically state that completion is synchronous.

vadimsemenov replied:
There may be a race with external future cancellation.

yorickhenning commented on integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:
```diff
@@ -421,42 +417,64 @@ private class OuterFuture<T>(private val deferred: Deferred<T>): ListenableFutur
 }
+
+/** Represents result of a `Coroutine` (either [Deferred] or [AbstractCoroutine]). */
+// This is similar to kotlin.Result type. However, kotlin.Result can't be used by external libraries.
```
It can't? It's an inline class in the core SDK. It's not a JVM-specific type.

vadimsemenov replied:
It was a surprise to me as well: https://stackoverflow.com/a/52632328

yorickhenning commented on integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:
```diff
@@ -394,17 +398,9 @@ private class OuterFuture<T>(private val deferred: Deferred<T>): ListenableFutur
-        }
+    private fun getInternal(result: Result<T>): T = when(result) {
+        is Result.Success -> result.value
+        is Result.Failure -> throw result.asFutureException()
```
This reallocates an Exception each time the Future is observed. Memoizing the exception is probably a good idea.

I hope no one cares about the reference equality of the Exceptions thrown, but it's not good to generate multiple distinct exception objects from a single failure.

yorickhenning replied:
Adding another patch or filing a bug are both reasonable plans here.

vadimsemenov replied:
That's true. 
I'll try to normalise the exception at the moment of coroutine completion.
It means that stack trace of `ExecutionException` would be somewhat messed. But I hope no one looks at the stack trace of `ExecutionException`...

vadimsemenov replied:
Done.

yorickhenning commented on integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:
```diff
@@ -421,42 +417,64 @@ private class OuterFuture<T>(private val deferred: Deferred<T>): ListenableFutur
-        } else {
-            false
+    @Suppress("UnstableApiUsage")
+    override fun afterDone() {
```
I vaguely remember there being a good reason not to use `afterDone()`. I'm pretty sure it came up last time around, maybe on an internal draft CR somewhere...

Seems wrong to override `cancel()` though. We should ask Chris.

Looks like I added test cases for `setFuture(cancelledFuture())`. Maybe there's some other completion-by-delegation corner case I've forgotten about.

vadimsemenov replied:
After our offline discussion, reverted back to override `cancel()` and added a TODO to reconsider this when afterDone API is finalised.

yorickhenning commented on integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:
```diff
@@ -421,42 +417,64 @@ private class OuterFuture<T>(private val deferred: Deferred<T>): ListenableFutur
+    class Failure<T>(val throwable: Throwable) : Result<T>()
+}
+
+/**
```
These extension functions would probably be clearer as members of the type declaration, given that it's right here and all. :)

vadimsemenov replied:
Done.

yorickhenning commented on integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:
```diff
@@ -421,42 +417,64 @@ private class OuterFuture<T>(private val deferred: Deferred<T>): ListenableFutur
-        set(deferred)
-    }
+    /**
+     * When entangled coroutine [completes][Job.isCompleted] (either successfully or exceptionally),
```
As above with the parenthetical.

Related, Kotlin's Job is _extremely_ specific about terms: https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-job/

Perhaps phrasing this as "When the attached coroutine [\`isCompleted\`][Job.isCompleted]" would be more specific?

vadimsemenov replied:
I agree. Reworded.

yorickhenning commented on integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:
````diff
@@ -70,10 +70,21 @@ public fun <T> CoroutineScope.future(
+ * deferred, **does not** cancel [Deferred] and original [ListenableFuture].
+ * To propagate cancellation from awaiting coroutine to the [ListenableFuture], use [ListenableFuture.await].
+ *
+ * ```
````
Why is this? The cancellation should be propagated no matter the adaption.

I added [this](https://github.com/vadimsemenov/kotlinx.coroutines/blame/c4b4e1c0211818ed3d01bdf988e7749cbb95f910/integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt#L384) deferred-as-LF-as-deferred test case last time around. Did I miss a case, or is one of the `await()` implementations not cancellation-correct?

vadimsemenov replied:
Cancellation of the Deferred will cancel original ListenableFuture, that's fine. However, cancellation of _the coroutine that calls `Deferred.await`_ doesn't cancel the Deferred.

This is kind of "by design" of the Deferred. Deferred represents a computation that can be `await`ed multiple times. So one can say that it's unfair to cancel the Deferred if a coroutine that `await`s on the Deferred is cancelled. Because other coroutines that `await` on the same Deferred may succeed. 

Here is a simple example: https://pl.kotl.in/DYcyjkku2
If you replace the `async` coroutine builder with `future`, future will be cancelled as expected. 
I think it's the main difference between `ListenableFuture.await` and `Deferred.await`. `ListenableFuture.await` is intended for one-shot future. `Deferred.await` assumes that the deferred can be reused.

You can also take a look at [awaitFutureWithCancel](https://github.com/Kotlin/kotlinx.coroutines/blob/master/integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt#L549) in tests.

yorickhenning replied:
OK, I'm following now.

I recommend striking the paragraph, then. The behaviour of `Deferred` isn't what this is documenting, it's documenting the adapter's behaviour.

Describing the cancellation of a calling coroutine and `await()` is adjacent to what this method is doing, and it confused me. I might need to care about that behaviour, but I should be discovering it through reading about `await()`, not here.

Some examples might be warranted, but I'm inclined to include positive examples rather than this one - cancellation works here and there, possibly contrasted with ordinary `Deferred` cancellation. It's also fair to leave those to the unit tests.

vadimsemenov replied:
Removed the paragraph. I think `await()` documents its own implementation good enough.

yorickhenning commented on integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:
````diff
@@ -70,10 +70,21 @@ public fun <T> CoroutineScope.future(
+ *
+ * ```
+ * suspend fun foo() {
+ *   listenableFuture.asDeferred().await()  // Doesn't propagate `foo`'s cancellation to the `listenableFuture`
````
We're in the weeds on semantics, but `foo` isn't cancellable, because it's a function. The example isn't quite complete without a coroutine invoking `foo`. Using a `launch` coroutine builder might be easier to write technically correct comments on, when explaining calling-coroutine cancellation.

vadimsemenov replied:
Fair enough. Replaced with `launch`.

yorickhenning commented on integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt:
```diff
@@ -540,12 +569,100 @@ class ListenableFutureTest : TestBase() {
     }
 
+    @Test
+    fun testFutureScopeCancellationPropagatedAlong() = runTest {
```
`testCancellingFutureContextJobCancelsFuture`

yorickhenning commented on integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt:
```diff
@@ -540,12 +569,100 @@ class ListenableFutureTest : TestBase() {
+        yield()
+        expect(2)
+        // Blocking get should succeed after internal coroutine completes.
+        assertFailsWith<ExecutionException> { future.get() }.run {
```
Use `getDone()` if the future is known to be complete.

vadimsemenov replied:
In this particular case it may not be done at this point. Future's coroutine is working in a different thread pool, so I indeed want to call a blocking get here. 
I didn't manage to find a better way to test that future isDone only after coroutine isCompleted. Can you see a better option?

yorickhenning replied:
Ah, I misunderstood. `get()` is the right way to block the test thread awaiting completion, then.

yorickhenning commented on integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt:
```diff
@@ -540,12 +569,100 @@ class ListenableFutureTest : TestBase() {
+        // Blocking get should succeed after internal coroutine completes.
+        assertFailsWith<ExecutionException> { future.get() }.run {
+            expect(4)
+            assertTrue(cause is TestException)
```
This is a moderately exciting use of `run {}`. Reading this requires the reader to know that `assertFailsWith` returns the `Exception`, and to mentally track several layers of implicit typing.

This'd be less functionally ambitious as:

```
val thrown = assertFailsWith<ExecutionException> { getDone(future) }
assertTrue(thrown.cause is TestException)
```

vadimsemenov replied:
Done.

yorickhenning commented on integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt:
```diff
@@ -540,12 +569,100 @@ class ListenableFutureTest : TestBase() {
+            foo.invokeOnCompletion {
+                expect(3)
+            }
+            val bar = async<Int> { throw TestException() }
```
To strengthen the `is` below, create a `TestException` `val` scoped to the `runTest {}` block, throw that object instance here, and use object equality to assert that the expected exception was thrown, rather than a type introspection.

It's not likely to be some other `TestException` in this case, but it's the stronger statement, and it's *really* asserting that the expected exception was thrown. It's important with `IllegalStateException` in more complex cases.

vadimsemenov replied:
Done.

yorickhenning commented on integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt:
```diff
@@ -315,6 +315,27 @@ class ListenableFutureTest : TestBase() {
     }
 
+    @Test
+    fun testAsListenableFuturePropagatesCancellationBeforeRunningListeners() = runTest {
```
Interesting...

Does this test case pass before the production code changes?

vadimsemenov replied:
This one didn't. 
That's the problem of overriding `AbstractFuture.cancel` and calling `super.cancel` before cancelling the deferred. Listeners are executed inside `AbstractFuture.cancel`.

vadimsemenov replied:
Similar testFuturePropagatesCancellationBeforeRunningListeners (for `CoroutineScope.future`) passed. But only because of AbstractFuture's implementation detail — it runs listeners in order of addition. And the listener that cancels the coroutine is the first one.

vadimsemenov committed:
Apply suggestions from the review

vadimsemenov committed:
Wrap coroutine's exception at the moment of coroutine completion

Now ExecutionException won't be reallocated on every get() call.

elizarov committed:
Removed unused NativeThreadLocal declaration

thibseisel committed:
Replace references to deprecated delayEach operator (#2127)

Co-authored-by: Thibault SEISEL <thibault.seisel@orange.com>

erokhins committed:
Update api dump for 1.4.0 compiler

elizarov committed:
Flow.transformWhile operator (#2066)

Also, most flow-truncating operators are refactored via a common internal collectWhile operator that properly uses AbortFlowException and checks for its ownership, so that we don't have to look for bugs in interactions between all those operators (and zip, too, which is also flow-truncating). But `take` operator still users a custom highly-tuned implementation.

Fixes #2065

Co-authored-by: EdwarDDay <4127904+EdwarDDay@users.noreply.github.com>
Co-authored-by: Louis CAD <louis.cognault@gmail.com>

qwwdfsad committed:
Replace scanReduce with runningReduce to be consistent with Kotlin standard library. (#2139)

foldReduce is not introduced to see how it goes in the standard library first. For the full rationale of renaming, please refer to KT-38060

elizarov committed:
Rename 'foo' function to 'simple' in flow docs (#2078)

The name 'foo' might be hard to get for novice developers. It is really hard to find a "meaningful" name for all examples in the docs, so this is simply a consistent solution. The alternative is to something like "simpleFlow", but adding a "Flow" suffix to the name implicitly condones Hungarian notation, which is not aligned with Kotlin coding style.

elizarov committed:
Fix race in Flow.asPublisher (#2124)

The race was leading to emitting more items via onNext than requested, the corresponding stress-test was added, too

Fixes #2109

elizarov committed:
Merge remote-tracking branch 'origin/master' into develop

elizarov committed:
Consistent toString for MainCoroutineDispatcher implementations (#2131)

So that is shows as "Dispatchers.Main" and "Dispatchers.Main.immediate".
Also remove hardcoded "Main" name in places of code where it is not needed anymore.

elizarov committed:
Increase timeout in the example to make sure it fails with threads (#2141)

Fixes #2135

elizarov committed:
Version 1.3.8

elizarov committed:
Coroutines debugger should keep weak references to running coroutines (#2129)

It should not prevent garbage-collection of coroutines that were otherwise lost, which included the following practically-useful cases:
* Synchronous coroutines (iterator/sequence).
* Lazy coroutines that were not started.
* Abandoned coroutines that suspend forever without strong references to them in GlobalScope.

Two kinds of tests cover this functionality:
* A test via FieldWalker ensures that debugger impl does not keep a strong reference. This tests works fast and provides good diagnostics if anything goes wrong, but it is fragile, as futures changes to debugger my introduce static references to running coroutines elsewhere.
* A stress-test that ensures that no OOM indeed happens when you run a lot of such lost coroutines. Longer-running, more stable to code change, but fragile in a difference sense as it may accidentally start passing in the future if lots of memory get allocated for tests.

Fixes #2117

elizarov committed:
Fix hanging RunInterruptibleStressTest on Windows with JDK 1.6 (#2145)

The test is improved so that it fails, not hangs, on a failure. However, it will not pass on Windows + JDK 1.6 due to a bug in JDK (which can be easily confirmed with this modification to a test), so it excluded when running on JDK 1.6.

Fixes #2144

1zaman committed:
Fix docs for CancellableContinuation (#2112)

Basically reverts the documentation update in commit
49b63053698195b645002ee0ad530a2c4e315178 due to the changes in commit
876e9bab95d34f311be030a0706497564f96d95a

OscarSpruit committed:
Add a keep rule for AndroidDispatcherFactory for R8 version > 1.6.0 (#2154)

In commit #3a8a0ea this keep rule was removed from kotlinx-coroutines-core (which is good), but
this caused AndroidDispatcherFactory to be minimized in newer R8 versions. Adding a rule explicitly
fixes the problem.

vadimsemenov committed:
Cancel coroutine in Future.cancel instead of AbstractFuture.afterDone

afterDone is still in beta. This also allowed to replace InnerFuture with SettableFuture.
Two tests are ignored because now cancellation propagates after listeners are executed. There's a TODO to fix this. Even though it doesn't violate any contract, it would be good to cancel the coroutine before executing the listeners.

vadimsemenov committed:
Remove confusing comment about difference in Deferred and ListenableFuture cancellation

yorickhenning commented on integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:
```diff
@@ -407,49 +405,25 @@ private class OuterFuture<T>(jobToCancel: Job): ListenableFuture<T> {
-        if (isCancelled) {
+    override fun cancel(mayInterruptIfRunning: Boolean): Boolean {
+        // TODO: call jobToCancel.cancel() _before_ running the listeners.
+        //  Future.cancel may execute all the listeners which can delay cancellation of jobToCancel.
```
"innerFuture.cancel() will execute innerFuture's listeners. This delays cancellation of jobToCancel until after innerFuture's listeners have already run." 

Something like that?

The distinction between "innerFuture" and "this Future" is a bit squishy. Maybe naming innerFuture "delegate" or something would be clearer. Not sure.

vadimsemenov replied:
Fixed the comment.

It's not exactly a "delegate", right? Renamed `innerFuture` to `auxFuture` and `OuterFuture` to `JobListenableFuture`. Let me know if you can think of a better naming.

yorickhenning replied:
No, I don't have a better name to hand. "innerFuture", "actualFuture" or "internalFuture", anything like that...

The inner Future is the concrete implementation, and the outer Future is a proxy type that both owns the reference and invokes methods on it. I haven't read the Gang of Four in long enough to describe this design pattern better.

yorickhenning commented on integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt:
```diff
@@ -316,6 +317,7 @@ class ListenableFutureTest : TestBase() {
     }
 
     @Test
+    @Ignore  // TODO: propagate cancellation before running listeners.
```
File and link an issue in the TODO(). If not with this PR, in a followup.

vadimsemenov replied:
Yep, will do.

yorickhenning commented on integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:
```diff
@@ -195,12 +196,18 @@ private fun ExecutionException.nonNullCause(): Throwable {
 public fun <T> Deferred<T>.asListenableFuture(): ListenableFuture<T> {
     val outerFuture = OuterFuture<T>(this)
-    outerFuture.afterInit()
+    // This invokeOnCompletion completes the innerFuture as `deferred` does. The innerFuture may
```
s/does/is completed/

Right?

vadimsemenov replied:
Reworded.

yorickhenning commented on integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:
```diff
@@ -415,48 +405,54 @@ private class OuterFuture<T>(private val deferred: Deferred<T>): ListenableFutur
 }
+
+/**
+ * Represents result of a `Coroutine` in the sense of [Future].
```
s/result/the result/

vadimsemenov replied:
Done.

yorickhenning commented on integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:
```diff
@@ -415,48 +405,54 @@ private class OuterFuture<T>(private val deferred: Deferred<T>): ListenableFutur
+}
+
+/**
+ * Wraps exception in a `Coroutine` into [Result.Failure]. [Result.Failure.exception] represents `this`
```
"Wraps an Exception thrown from a coroutine into" perhaps.

vadimsemenov replied:
Done. Emphasised that the exception is the receiver.

yorickhenning reviewed (approved):
Looks good! Probably time to send it upstream.

vadimsemenov committed:
Amend comments as Yorick suggested

Also rename OuterFuture to JobListenableFuture and innerFuture to auxFuture.

charlesmunger commented on integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:
```diff
@@ -333,130 +327,133 @@ private class ListenableFutureCoroutine<T>(
-        return isDone
-          && (innerFuture.isCancelled
-          || deferred.getCompletionExceptionOrNull() is kotlinx.coroutines.CancellationException)
+        return isDone &&
```
You can avoid the isDone check in some cases by doing it *after* you check isCancelled

vadimsemenov replied:
Done.

charlesmunger commented on integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:
```diff
@@ -333,130 +327,133 @@ private class ListenableFutureCoroutine<T>(
+ */
+private fun <T> Throwable.asResult() = Result.Failure<T>(
+    if (this is CancellationException) this
+    else ExecutionException(this)
```
Both CancellationException and ExecutionException will have the wrong stack trace here. You can instead change result to be a sealed type with Success and Cancellation - use settablefuture.setException() for failures. 

This would give you settablefuture's ExecutionException wrapping for free; you'd create a new CancellationException in get(), with the input one as the cause.

vadimsemenov replied:
Thanks, done.

charlesmunger commented on integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:
```diff
@@ -333,130 +327,133 @@ private class ListenableFutureCoroutine<T>(
  */
-private class OuterFuture<T>(private val deferred: Deferred<T>): ListenableFuture<T> {
-    val innerFuture = DeferredListenableFuture(deferred)
+private class JobListenableFuture<T>(private val jobToCancel: Job): ListenableFuture<T> {
```
Propagate toString() as done in AbstractFuture

MasoodFallahpoor committed:
Add missing article (#2180)

elizarov committed:
Support context in Flow.asPublisher and similar methods (#2156)


Fixes #2155

Co-authored-by: Vsevolod Tolstopyatov <qwwdfsad@gmail.com>

katia-energizer committed:
Feat: updated debugging coroutines in the documentation (#2191)

qwwdfsad committed:
Version 1.3.9

qwwdfsad committed:
Update documentation

MasoodFallahpoor committed:
Update flow.md (#2196)

shubham08gupta committed:
Update typo in flow.md (#2199)

jsoendermann committed:
Add missing 'the' (#2200)

qwwdfsad committed:
Merge pull request #2203 from Kotlin/version-1.3.9

Version 1.3.9

qwwdfsad committed:
Remove redundant release step, it has been automated by try.kotlinlang.org team

qwwdfsad committed:
Remove redundant and flaky (due to its nature, not implementation problems) DebugLeaksStressTest

qwwdfsad committed:
Fix sporadic ConcurrentWeakMapTest failures (#2187)

qwwdfsad committed:
Cherry-picks from native-mt branch to reduce maintenance burden (#2204)

* Immediate dispatcher on JS, test added
* Style fixes

turansky committed:
Kotlin and Slack channel shields (#2205)

qwwdfsad committed:
Add space staging repository

qwwdfsad committed:
Migrate examples to new JS plugin

    * General cleanup and TODOs

Co-authored with: ilya.goncharov@jetbrains.com

turansky committed:
New Kotlin/JVM & Kotlin/Multiplatform plugins (#2018)

vadimsemenov committed:
Refactor isCancelled to save a volatile read.

vadimsemenov committed:
Refactor JobListenableFuture so ExecutionException is handled by SettableFuture.

Also propagate CancellationException from coroutines as a cause of new CancellationException. So outer CancellationException has correct stack trace.

vadimsemenov committed:
Implement JobListenableFuture.toString in a similar to AbstractFuture.toString fashion.

AbstractFuture.toString has some useful info about waiters and existing timeouts.
